zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs

```diff
@@ -0,0 +1,384 @@
+            cb.call_context_lookup(false.expr(), None, CallContextFieldTag::TxId, tx_id.expr());
+
+            // Check if EXTCODECOPY external address is warm.
+            cb.account_access_list_read(
+                tx_id.expr(),
+                from_bytes::expr(&external_address.cells[..N_BYTES_ACCOUNT_ADDRESS]),
+                is_warm.expr(),
+            );
+
+            // EXTCODECOPY has an extra stack pop for external address.
+            cb.stack_pop(external_address.expr());
+        });
+
+        cb.stack_pop(dst_offset.expr());
+        cb.stack_pop(src_offset.expr());
+        cb.stack_pop(copy_size.expr());
+
+        let dst_memory_addr = MemoryAddressGadget::construct(cb, dst_offset, copy_size);
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [dst_memory_addr.address()]);
+        let memory_copier_gas = MemoryCopierGasGadget::construct(
+            cb,
+            dst_memory_addr.length(),
+            memory_expansion.gas_cost(),
+        );
+
+        let constant_gas_cost = select::expr(
+            is_extcodecopy.expr(),
+            // According to EIP-2929, EXTCODECOPY constant gas cost is different for cold and warm
+            // accounts.
+            select::expr(
+                is_warm.expr(),
+                GasCost::WARM_ACCESS.expr(),
+                GasCost::COLD_ACCOUNT_ACCESS.expr(),
+            ),
+            // Constant gas cost is same for CALLDATACOPY, CODECOPY and RETURNDATACOPY.
+            OpcodeId::CALLDATACOPY.constant_gas_cost().expr(),
+        );
+
+        let insufficient_gas = LtGadget::construct(
+            cb,
+            cb.curr.state.gas_left.expr(),
+            constant_gas_cost + memory_copier_gas.gas_cost(),
+        );
+
+        cb.require_equal(
+            "Gas left is less than gas cost",
+            insufficient_gas.expr(),
+            1.expr(),
+        );
+
+        let common_error_gadget = CommonErrorGadget::construct(
+            cb,
+            opcode.expr(),
+            // EXTCODECOPY has extra 1 call context lookup (tx_id), 1 account access list
+            // read (is_warm), and 1 stack pop (external_address).
+            5.expr() + 3.expr() * is_extcodecopy.expr(),
+        );
+
+        Self {
+            opcode,
+            is_warm,
+            tx_id,
+            external_address,
+            src_offset,
+            dst_memory_addr,
+            memory_expansion,
+            memory_copier_gas,
+            insufficient_gas,
+            is_extcodecopy,
+            common_error_gadget,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        transaction: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        let is_extcodecopy = usize::from(opcode == OpcodeId::EXTCODECOPY);
```

Comment:
i would prefer keep is_extcodecopy as boolean and write `let rw_offset = if is_extcodecopy { 3 } else { 0 };` below.

Rust is not C

Reply:
Fixed in commit [9410eda](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1288/commits/9410eda35dd11f9f8c386aa562dd2c996b461638).